src/quantization/sq4_uniform_quantizer.h

```diff
@@ -0,0 +1,247 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <math.h>
+
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "quantizer.h"
+#include "simd/sq4_uniform_simd.h"
+
+namespace vsag {
+
+using norm_type = uint64_t;
+using sum_type = float;
+
+template <MetricType metric = MetricType::METRIC_TYPE_L2SQR>
+class SQ4UniformQuantizer : public Quantizer<SQ4UniformQuantizer<metric>> {
+public:
+    explicit SQ4UniformQuantizer(int dim);
+
+    bool
+    TrainImpl(const DataType* data, uint64_t count);
+
+    bool
+    EncodeOneImpl(const DataType* data, uint8_t* codes) const;
+
+    bool
+    EncodeBatchImpl(const DataType* data, uint8_t* codes, uint64_t count);
+
+    bool
+    DecodeOneImpl(const uint8_t* codes, DataType* data);
+
+    bool
+    DecodeBatchImpl(const uint8_t* codes, DataType* data, uint64_t count);
+
+    inline float
+    ComputeImpl(const uint8_t* codes1, const uint8_t* codes2) const;
+
+    inline void
+    ProcessQueryImpl(const DataType* query, Computer<SQ4UniformQuantizer>& computer) const;
+
+    inline void
+    ComputeDistImpl(Computer<SQ4UniformQuantizer>& computer,
+                    const uint8_t* codes,
+                    float* dists) const;
+
+private:
+    DataType lower_bound_{0};
+    DataType diff_{0};
+    uint64_t offset_code_{0};
+    uint64_t offset_norm_{0};
+    uint64_t offset_sum_{0};
+};
+
+template <MetricType metric>
+SQ4UniformQuantizer<metric>::SQ4UniformQuantizer(int dim)
+    : Quantizer<SQ4UniformQuantizer<metric>>(dim) {
+    lower_bound_ = std::numeric_limits<DataType>::max();
+    diff_ = std::numeric_limits<DataType>::min();
+
+    this->code_size_ = 0;
+
+    offset_code_ = this->code_size_;
+    this->code_size_ += (dim + 1) >> 1 << 1;
+
+    if (metric == MetricType::METRIC_TYPE_L2SQR or metric == MetricType::METRIC_TYPE_COSINE) {
+        offset_norm_ = this->code_size_;
+        this->code_size_ += sizeof(norm_type);  // norm of vector
+    }
+
+    if (metric == MetricType::METRIC_TYPE_IP or metric == MetricType::METRIC_TYPE_COSINE) {
+        offset_sum_ = this->code_size_;
+        this->code_size_ += sizeof(sum_type);  // sum of vector
+    }
+}
+
+template <MetricType metric>
+bool
+SQ4UniformQuantizer<metric>::TrainImpl(const DataType* data, uint64_t count) {
+    if (data == nullptr) {
+        return false;
+    }
+
+    for (uint64_t i = 0; i < count; i++) {
+        for (uint64_t d = 0; d < this->dim_; d++) {
+            auto val = data[i * this->dim_ + d];
+            if (val > diff_) {
+                diff_ = val;
+            }
+            if (val < lower_bound_) {
+                lower_bound_ = val;
+            }
+        }
+    }
+
+    diff_ -= lower_bound_;
+    if (diff_ < 1e-4) {
+        diff_ = 1;
+    }
```

Comment:
why diff_ = 1

Reply:
actually, when diff_ < 1e-4, we should throw a warring. In such situation, the quantizer is meaningless

Reply:
add todo